### PR TITLE
Add ssh-agent support

### PR DESCRIPTION
**Pull request by arturrez**
resolves https://github.com/ava-labs/avalanche-cli/issues/1298
adds `--use-ssh-agent` to read public key from `ssh-agent` for the identity provided via `--ssh-identity`. Or prompts for one. It automatically detects Yibikey based on `cardno` comment and marks it as recommended.
Nodes created with this will have empty private key set in inventory and `cli` will use agent to connect to them going forward. 
usage example 
`avalanche node create test2  --use-static-ip=false --authorize-access=true --latest-avalanchego-version=true --aws --aws-profile=ddefault  --region us-east-1 --num-nodes=1 --fuji --use-ssh-agent --ssh-identity=cardno:XX_XXX_XXX`

### CODE REVIEW

**Comment by arturrez** [comment]
> Need to update the ssh instruction, other than that lgtm

do you mean avalanche-docs?

File: cmd/nodecmd/create.go [hunk 1]
```diff
@@ -120,6 +125,12 @@ func preCreateChecks() error {
 			}
 		}
 	}
+	if len(sshIdentity) > 0 && !useSSHAgent {
```
Comment (felipemadero) [inline cmd/nodecmd/create.go:128]: better to compare with "" ?

File: cmd/nodecmd/create.go [hunk 2]
```diff
@@ -97,6 +100,8 @@ will apply to all nodes in the cluster`,
 	cmd.Flags().StringVar(&awsProfile, "aws-profile", constants.AWSDefaultCredential, "aws profile to use")
 	cmd.Flags().BoolVar(&createOnFuji, "fuji", false, "create node/s in Fuji Network")
 	cmd.Flags().BoolVar(&createDevnet, "devnet", false, "create node/s into a new Devnet")
+	cmd.Flags().BoolVar(&useSSHAgent, "use-ssh-agent", false, "use ssh agent(ex: Yubikey) for ssh auth")
+	cmd.Flags().StringVar(&sshIdentity, "ssh-identity", "", "use given ssh identity")
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/nodecmd/create.go:104]: maybe call it ssh-agent-identity

File: cmd/nodecmd/create.go [hunk 3]
```diff
@@ -782,3 +800,25 @@ func getRegionsNodeNum(cloudName string) (
...
 }
+
+func setSSHIdentity() (string, error) {
+	const yubikeyMark = " [YubiKey] (recommended)"
+	sshIdentities, err := utils.ListSSHIdentity()
+	if err != nil {
+		return "", err
+	}
+	sshIdentities = utils.Map(sshIdentities, func(id string) string {
+		yubikeyPattern := `cardno:(\d+(_\d+)*)`
```
Comment (felipemadero) [inline cmd/nodecmd/create.go:811]: this can be a constant

File: cmd/nodecmd/create.go [hunk 4]
```diff
@@ -782,3 +800,25 @@ func getRegionsNodeNum(cloudName string) (
...
+		if len(regexp.MustCompile(yubikeyPattern).FindStringSubmatch(id)) > 0 {
```
Comment (felipemadero) [inline cmd/nodecmd/create.go:812]: this may be compiled before the map

File: cmd/nodecmd/create_aws.go [hunk 5]
```diff
@@ -139,41 +139,54 @@ func createEC2Instances(ec2Svc map[string]*awsAPI.AwsCloud,
...
 		if err != nil {
 			return nil, nil, nil, nil, err
 		}
-		certName := regionConf[region].CertName
+		sgID := ""
 		keyPairName[region] = regionConf[region].Prefix
 		securityGroupName := regionConf[region].SecurityGroupName
-		privKey, err := app.GetSSHCertFilePath(certName)
-		sgID := ""
+		privKey, err := app.GetSSHCertFilePath(regionConf[region].CertName)
```
Comment (felipemadero) [inline cmd/nodecmd/create_aws.go:151]: maybe for readability we need two function to replace here, one for the ssh agent case, and the other for the ssh file case

File: cmd/nodecmd/create.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/nodecmd/create.go:104]: fixed

File: cmd/nodecmd/create.go (hunk 1 above)
Comment (arturrez) [reply cmd/nodecmd/create.go:128]: fixed

File: cmd/nodecmd/create.go (hunk 3 above)
Comment (arturrez) [reply cmd/nodecmd/create.go:811]: fixed

File: cmd/nodecmd/create.go (hunk 4 above)
Comment (arturrez) [reply cmd/nodecmd/create.go:812]: good catch. fixed.

File: cmd/nodecmd/create_aws.go (hunk 5 above)
Comment (arturrez) [reply cmd/nodecmd/create_aws.go:151]: not sure wdym. I suggest leaving this as is unless you want to open a PR with your implementation

Comment (arturrez) [reply cmd/nodecmd/create_aws.go:151]: i think you want to fix this massibe if/switch statement combo.
we should do it, but maybe in separate PR

File: cmd/nodecmd/wiz.go [hunk 6]
```diff
@@ -85,6 +85,8 @@ The node wiz command creates a devnet and deploys, sync and validate a subnet in
 	cmd.Flags().StringVar(&nodeConf, "node-config", "", "path to avalanchego node configuration for subnet")
 	cmd.Flags().StringVar(&subnetConf, "subnet-config", "", "path to the subnet configuration for subnet")
 	cmd.Flags().StringVar(&chainConf, "chain-config", "", "path to the chain configuration for subnet")
+	cmd.Flags().BoolVar(&useSSHAgent, "use-ssh-agent", false, "use ssh agent for ssh")
+	cmd.Flags().StringVar(&sshIdentity, "ssh-identity", "", "use given ssh identity")
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/nodecmd/wiz.go:89]: fixed

File: pkg/ansible/ansible.go [hunk 7]
```diff
@@ -39,7 +39,7 @@ func CreateAnsibleHostInventory(inventoryDirPath string, cloudConfigMap models.C
 			inventoryContent += publicIPMap[instanceID]
 			inventoryContent += " ansible_user=ubuntu"
 			inventoryContent += fmt.Sprintf(" ansible_ssh_private_key_file=%s", cloudConfig.CertFilePath)
-			inventoryContent += fmt.Sprintf(" ansible_ssh_common_args='%s'", constants.AnsibleSSHInventoryParams)
+			inventoryContent += fmt.Sprintf(" ansible_ssh_common_args='%s'", constants.AnsibleSSHUseAgentParams)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/ansible/ansible.go:42]: it adds `"-o StrictHostKeyChecking=no"` which is needed to make ssh skip verification against local public key file list

File: pkg/cloud/aws/aws.go [hunk 8]
```diff
@@ -336,6 +336,23 @@ func (c *AwsCloud) CreateAndDownloadKeyPair(keyName string, privateKeyFilePath s
...
+	if err != nil {
+		return err
+	}
+	if _, err = c.ec2Client.ImportKeyPair(c.ctx, &ec2.ImportKeyPairInput{
+		KeyName:           aws.String(keyName),
+		PublicKeyMaterial: []byte(publicKeyMaterial),
+	}); err != nil {
+		return err
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/cloud/aws/aws.go:353]: agree

File: pkg/cloud/aws/aws.go [hunk 9]
```diff
@@ -336,6 +336,23 @@ func (c *AwsCloud) CreateAndDownloadKeyPair(keyName string, privateKeyFilePath s
 	return nil
 }
 
+func (c *AwsCloud) UploadKeyPair(keyName string, identity string) error {
+	if !utils.IsSSHIdentityValid(identity) {
+		return fmt.Errorf("identity %s not found", identity)
+	}
+	publicKeyMaterial, err := utils.ReadSSHIdentityPublicKey(identity)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/cloud/aws/aws.go:343]: this is ssh public key, so name is correct

File: pkg/cloud/aws/aws.go [hunk 10]
```diff
@@ -336,6 +336,23 @@ func (c *AwsCloud) CreateAndDownloadKeyPair(keyName string, privateKeyFilePath s
 	return nil
 }
 
+func (c *AwsCloud) UploadKeyPair(keyName string, identity string) error {
+	if !utils.IsSSHIdentityValid(identity) {
+		return fmt.Errorf("identity %s not found", identity)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/cloud/aws/aws.go:341]: fixed

File: pkg/models/host.go [hunk 11]
```diff
@@ -36,9 +36,18 @@ type Host struct {
 }
 
 func NewHostConnection(h *Host) (*goph.Client, error) {
-	auth, err := goph.Key(h.SSHPrivateKeyPath, "")
-	if err != nil {
-		return nil, err
+	var auth goph.Auth
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/models/host.go:39]: fixed

File: pkg/models/host.go [hunk 12]
```diff
@@ -36,9 +36,18 @@ type Host struct {
...
+	var err error
+	if h.SSHPrivateKeyPath == "" {
+		auth, err = goph.UseAgent()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		auth, err = goph.Key(h.SSHPrivateKeyPath, "")
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/models/host.go:48]: i like it

File: pkg/utils/ssh.go [hunk 13]
```diff
@@ -4,10 +4,81 @@ package utils
...
+	if certFilePath != "" {
+		return fmt.Sprintf("ssh %s %s@%s -i %s", constants.AnsibleSSHShellParams, constants.AnsibleSSHUser, publicIP, certFilePath)
+	} else {
+		return fmt.Sprintf("ssh %s %s@%s", constants.AnsibleSSHUseAgentParams, constants.AnsibleSSHUser, publicIP)
+	}
+}
+
+// isSSHAgentAvailable checks if the SSH agent is available.
+func IsSSHAgentAvailable() bool {
+	return os.Getenv("SSH_AUTH_SOCK") != ""
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:27]: i think no need for this at least for now

File: pkg/utils/ssh.go [hunk 14]
```diff
@@ -4,10 +4,81 @@ package utils
...
+}
+
+func GetSSHAgent() (agent.ExtendedAgent, error) {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:30]: fixed

File: pkg/utils/ssh.go [hunk 15]
```diff
@@ -4,10 +4,81 @@ package utils
...
+	if !IsSSHAgentAvailable() {
+		return nil, fmt.Errorf("SSH agent is not available")
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:32]: no need imho

File: pkg/utils/ssh.go [hunk 16]
```diff
@@ -4,10 +4,81 @@ package utils
...
+	}
+	sshAuthSock := os.Getenv("SSH_AUTH_SOCK")
+	conn, err := net.Dial("unix", sshAuthSock)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:37]: agree

File: pkg/utils/ssh.go [hunk 17]
```diff
@@ -4,10 +4,81 @@ package utils
...
+	}
+	sshIDs, err := sshAgent.List()
+	if err != nil {
+		return nil, err
+	}
+	identityList := Map(sshIDs, func(id *agent.Key) string { return id.Comment })
+	return identityList, nil
+}
+
+func IsSSHIdentityValid(identity string) bool {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:57]: fixed

File: pkg/utils/ssh.go [hunk 18]
```diff
@@ -4,10 +4,81 @@ package utils
...
+	identityList, err := ListSSHIdentity()
+	if err != nil {
+		return false
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:60]: agree, but makes code a little bit more messy. let me fix it

File: pkg/utils/ssh.go [hunk 19]
```diff
@@ -4,10 +4,81 @@ package utils
...
+	}
+	return slices.Contains(identityList, identity)
+}
+
+func ReadSSHIdentityPublicKey(identityName string) (string, error) {
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:65]: sure

File: pkg/utils/ssh.go [hunk 20]
```diff
@@ -4,10 +4,81 @@ package utils
...
+	sshAuthSock := os.Getenv("SSH_AUTH_SOCK")
+	conn, err := net.Dial("unix", sshAuthSock)
+	if err != nil {
+		return nil, err
+	}
+	sshAgent := agent.NewClient(conn)
+	return sshAgent, nil
+}
+
+// ListSSHIdentity returns a list of SSH identities and any error encountered.
```
[the remark this answers is not in the document]
Comment (arturrez) [reply pkg/utils/ssh.go:43]: agree with everything. fixed

**Review by felipemadero — APPROVED** [review]
LGTM

**Pull request merged**